PayItForward.Data/Repositories/EfGenericRepository.cs

```diff
@@ -0,0 +1,144 @@
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Microsoft.EntityFrameworkCore;
+    using PayItForward.Data.Models;
+    using PayItForward.Data.Models.Abstraction;
+
+    public class EfGenericRepository<T, TKey> : IRepository<T, TKey>
+        where T : BaseModel<TKey>
+    {
+        public EfGenericRepository(PayItForwardDbContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentException("An instance of DbContext is required to use this repository.", "context");
+            }
+
+            this.Context = context;
+            this.DbSet = this.Context.Set<T>();
+        }
+
+        protected DbSet<T> DbSet { get; set; }
+
+        protected PayItForwardDbContext Context { get; set; }
+
+        public async Task<T> GetByIdAsync(TKey id)
+        {
+            return await this.DbSet.FindAsync(id);
+        }
+
+        public virtual T GetById(TKey id)
+        {
+            return this.DbSet.Find(id);
+        }
+
+        public void Add(T entity)
+        {
+            this.ChangeEntityState(entity, EntityState.Added);
+        }
+
+        public void Update(T entity)
+        {
+            var entry = this.Context.Entry(entity);
+            if (entry.State == EntityState.Detached)
+            {
+                this.DbSet.Attach(entity);
+            }
+
+            entry.State = EntityState.Modified;
+        }
+
+        public IQueryable<T> GetAll()
+        {
+            return this.DbSet;
+        }
+
+        public Task<IEnumerable<T>> GetAllAsync()
+        {
+            return Task.FromResult(this.DbSet.AsEnumerable());
+        }
+
+        // to be tested
+        public async Task UpdateAsync(T entity)
+        {
+            this.Context.Set<T>().Update(entity);
+            await this.Context.SaveChangesAsync();
+        }
+
+        public void SoftDelete(T userTodelete)
+        {
+            userTodelete.IsDeleted = true;
+            userTodelete.DeletedOn = DateTime.Now;
+            this.Update(userTodelete);
+        }
+
+        public async Task<int> HardDeleteAsync(T entity)
+        {
+            this.Context.Set<T>().Remove(entity);
+            return await this.Context.SaveChangesAsync();
+        }
+
+        public Task SaveAsync()
+        {
+            this.ApplyAuditInfoRules();
+            this.ApplyDeletableEntityRules();
+            return this.Context.SaveChangesAsync();
+        }
+
+        public void Dispose()
+        {
+            this.Context.Dispose();
+        }
+
+        private void ChangeEntityState(T entity, EntityState entityState)
+        {
+            // this.Context.Entry(entity).State = entityState;
+            var entry = this.Context.Entry(entity);
+
+            if (entry.State == EntityState.Detached)
+            {
+                this.DbSet.Attach(entity);
+            }
+
+            entry.State = entityState;
+        }
+
+        private void ApplyAuditInfoRules()
```

Comment:
You are applying audid info rules on both places for some of them, please review this